Add mechanism to avoid forced child selection on blocks with templates.

jorgefilipecosta opened the pull request:
## Description
Currently, if a block contains a template when we insert that block the block that gets focused is one of the child blocks specified in the template. We don't offer a way to disable this behavior.

This current behavior makes sense in most cases (e.g: columns) but for some cases, 
we may want to keep the focus on the parent. E.g. In PR https://github.com/WordPress/gutenberg/pull/9416 @afercia pointed some a11y concerns in automatically selecting the child block on the InnerBlocks area.

Blocks may have their own edition area and an InnerBlocks are at the end. Automatically selecting a block in the InnerBlocks area, has some accessible concerns has the parent block edition area may be unnoticeable for screen reader users.

This PR adds a flag to the InnerBlocks component that allows blocks to disable the behavior of automatically selecting the children. In order for this flag to be possible another flag was added in insertBlock(s) action that allows for blocks to be inserted without a selection update happening.

## How has this been tested?
I created some tests blocks available in https://gist.github.com/jorgefilipecosta/a99dcbaedd35ec88c37b142679bc0dc1.
I checked that when I insert 'Test No Template Selection' the parent block is the block selected after the insert.
I checked that when I insert 'Test Template Selection' the first child is the block that gets inserted.
Columns continue to work exactly as before (the first paragraph gets selected) when a column is inserted.

jorgefilipecosta committed:
Add mechanism to avoid forced child selection on blocks with templates.

youknowriad commented on packages/editor/src/store/reducer.js:
```diff
@@ -651,14 +651,18 @@ export function blockSelection( state = {
-				isMultiSelecting: false,
-			};
+		case 'INSERT_BLOCKS': {
+			if ( action.updateSelection ) {
```
For me, this is not what's causing the innerBlocks to be selected, it's the `focusTabbable` function inside `block.js`. This on ensures the inserted block (parent) is selected.

jorgefilipecosta replied:
But the children are inserted after the parent, so without this changes, the selection gets updated on the child blocks are inserted.

youknowriad replied:
Oh makes sense. Seems like a decent approach.

jorgefilipecosta replied:
Hi @youknowriad thank you for looking into this PR. Do you think it is ready to merge?

chrisvanpatten commented:
OH MY GOSH this is so exciting. 😁🤗

In my custom Columns block we dispatch an `insertBlock` action to [insert blocks from our own UI](https://github.com/TomodomoCo/gutenberg-block-columns/blob/master/src/block/inspector.js#L66) and it works but causes the paragraph inside the column to get selected. Being able to disable this behavior would be just the most incredible thing.

💟

youknowriad commented on docs/data/data-core-editor.md:
```diff
@@ -1514,7 +1514,10 @@ inserted, optionally at a specific index respective a root block list.
  * index: Index at which block should be inserted.
  * rootClientId: Optional root client ID of block list on which
-                              to insert.
+                                   to insert.
```
Maybe we should just not add a line break at all for arguments with long lines (to avoid messing with the markdown output)

youknowriad commented on packages/editor/src/components/inner-blocks/README.md:
```diff
@@ -92,6 +92,13 @@ const TEMPLATE = [ [ 'core/columns', {}, [
 
 The previous example creates an InnerBlocks area containing two columns one with an image and the other with a paragraph.
 
+### `templateInsertUpdatesSelection`
```
I wonder if this is really needed or if it should always be "false"?

jorgefilipecosta replied:
I agree for some blocks it may make sense to use a different setting.

youknowriad reviewed (commented):
Looks good to me overall. @aduth Thoughts on this new argument?

afercia commented:
Right now, the "Media & Text" block accessibility is less than ideal, see https://github.com/WordPress/gutenberg/pull/9416#issuecomment-429952703

In that PR it was mentioned there are higher order limitations with templates / nested blocks that, at that time, prevented to fully address the accessibility concerns. This PR is meant to solve one of those issues. However, there's no milestone set and the "Media & Text" block is still not accessible. We're now at less than a week from the scheduled Release Candidate.

Are there any plans to help this PR move on? /Cc @mtias @youknowriad @jorgefilipecosta @tofumatt 

Adding the label "bug" as "Media & Text" is not accessible.

jorgefilipecosta commented:
I rebased this PR and retested and things looked fine, it should be ready, I added it to 4.3 milestone lets we manage to get it to the finish line on time.

chrisvanpatten reviewed (approved):
I'm not comfortable commenting on the specific implementation, but I tested the feature with our custom columns block and it works _beautifully_. The user experience is dramatically improved by this.

Hope to see it make it into 4.3 :)

youknowriad reviewed (approved):
LGTM 👍
